[BasicAA] Check for Overflow using vscale_range

davemgreen opened the pull request:
This extends #80818 when IsNSW is lost (possibly due to looking through multiple GEPs), to check the vscale_range for an access that will not overflow even with the maximum range.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1183,12 +1183,22 @@ AliasResult BasicAAResult::aliasGEP(
+      // so noalias still holds so long as the dependency distance is at least
+      // as big as the typesize.
+      if (VLeftSize.hasValue() &&
+          Scale.uge(VLeftSize.getValue().getKnownMinValue()))
```
I missed this when reviewing the original patch: Isn't this Scale supposed to be abs(Scale)? I don't think this does the right thing for negative scales right now.

davemgreen replied:
Oh Yep. I'll put a patch together with a test.

davemgreen replied:
I left the Scale in this patch as it is used in a signed multiply. Let me know if you think that should change.

nikic reviewed (approved):
LGTM

davemgreen committed:
[BasicAA] Check for Overflow using vscale_range

This extends #80818 when IsNSW is lost (possibly due to looking through
multiple GEPs), to check the vscale_range for an access that will not overflow
even with the maximum range.
